zrml/prediction-markets/src/lib.rs

```diff
@@ -1349,6 +1350,338 @@ mod pallet {
+        /// if the `CloseMarketsEarlyOrigin` decides to reject
+        /// the early close request of the market creator or
+        /// if the `CloseMarketsEarlyOrigin` is inactive.
+        /// It is slashed, if the `CloseMarketsEarlyOrigin` decides to schedule the early close.
+        ///
+        /// # Weight
+        ///
+        /// Complexity: `O(n)`, where `n` is the maximum number of market ids
+        /// in `MarketIdsPerClose...` either at the old period end or new period end.
+        #[pallet::call_index(19)]
+        #[pallet::weight(T::WeightInfo::dispute_early_close(CacheSize::get(), CacheSize::get()))]
+        #[transactional]
+        pub fn dispute_early_close(
+            origin: OriginFor<T>,
+            #[pallet::compact] market_id: MarketIdOf<T>,
+        ) -> DispatchResultWithPostInfo {
+            let who = ensure_signed(origin)?;
+            let market = <zrml_market_commons::Pallet<T>>::market(&market_id)?;
+            Self::ensure_market_is_active(&market)?;
+            let mut early_close = market.early_close.ok_or(Error::<T>::NoEarlyCloseScheduled)?;
+            match early_close.state {
+                EarlyCloseState::ScheduledAsMarketCreator => (),
+                EarlyCloseState::ScheduledAsOther
+                | EarlyCloseState::Disputed
+                | EarlyCloseState::Rejected => {
+                    return Err(Error::<T>::InvalidEarlyCloseState.into());
+                }
+            };
+
+            // ensure we don't dispute if the old market period is already over
+            // so that we don't switch back to an invalid market period
+            // this should never trigger, because at the time of scheduling a new market period,
+            // we ensure the new end is always before the old end
+            debug_assert!(
+                {
+                    match early_close.old {
+                        MarketPeriod::Block(ref range) => {
+                            let now_block = <frame_system::Pallet<T>>::block_number();
+                            now_block < range.end
+                        }
+                        MarketPeriod::Timestamp(ref range) => {
+                            let now_time = <zrml_market_commons::Pallet<T>>::now();
+                            now_time < range.end
+                        }
+                    }
+                },
+                "This will never happen, because schedule_early_close ensures that the new end is \
+                 always before the old end, otherwise the switch to an old market period would \
+                 lead to a never ending market!"
+            );
+
+            let close_dispute_bond = T::CloseEarlyDisputeBond::get();
+
+            T::AssetManager::reserve_named(
+                &Self::reserve_id(),
+                Asset::Ztg,
+                &who,
+                close_dispute_bond,
+            )?;
+
+            let ids_len_0 = Self::clear_auto_close(&market_id)?;
```

Comment:
It seems to me that this could be used to mock the market creator. Example:
Market creator schedules a truthful close in let's say 10 days.
`CloseEarlyBlockPeriod=5days` passes.
shortly before the close would happen, someome disputes and effectively denies the truthful close.

I think we have to mitigate this by enforcing a dispute duration after a close request that is bound to `CloseEarlyBlockPeriod`. Of course, the privileged origin

Reply:
I don't understand this. A market creator directly schedules a close at `now + CloseEarlyBlockPeriod`. If someone disputes shortly before the close, indeed it would extend the closing time, because the AC has to decide from then on. But the disputant reserved a bond, which is slashed, if the market creator was right.

Reply:
Please verify if the following scenario could occur given the current implementation:
1. `CloseEarlyBlockPeriod = 100 blocks`
2. `creator` requests `close` in 500 blocks
3. `close` in 500 blocks is approved
4. 490 blocks later, malicious disputor disputes the close request, effectively postponing it
5. malicious disputor has insight of the outcome at 500 blocks (lookup in internet or something), but since the malicious disputor postponed the closing, he is able to buy all correct outcome tokens in the market
6. malicious disputor is in a profit, even when losing the close early dispute bond.

Reply:
> 1. `CloseEarlyBlockPeriod = 100 blocks`
> 2. `creator` requests `close` in 500 blocks

The creator can't request a close in 500 blocks, but only at `now + CloseEarlyBlockPeriod`.

> 3. `close` in 500 blocks is approved

This is conflicting with the point 4 you made. The market creators request is not approved if it is disputed.

> 4. 490 blocks later, malicious disputor disputes the close request, effectively postponing it

Yeah that's correct, the disputant postpones the close early request, effectively waiting for the AC to schedule a shorter early close or a rejection.

> 5. malicious disputor has insight of the outcome at 500 blocks (lookup in internet or something), but since the malicious disputor postponed the closing, he is able to buy all correct outcome tokens in the market

Good thought and now I see your point! The implemented feature is for the purpose of giving the market participants the opportunity to `redeem_shares` earlier than they would have done normally. It is a better user experience to not sit in your positions any longer than required. The market creator will only close at the point when the true outcome is already known anyways. So your described scenario could already play out before the market creator even thinks about closing the market early.

> 6. malicious disputor is in a profit, even when losing the close early dispute bond.

The profit at this point is much lower than the expected profit at the point of the true outcome becoming known. Assuming the true outcome is below 1.

Reply:
I see, I thought the purpose was to correct the closing date based on the availability of information of the outcome.